### PR TITLE
[Kernel][Model] logits_soft_cap for Gemma2 with flashinfer

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**
Add logits_soft_cap for flashinfer, which is needed by Gemma2 model, also add a simple gemma2 test.

### CODE REVIEW

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -659,6 +659,16 @@ def _prepare_model_input_tensors(
...
 
+        logits_soft_cap = getattr(self.model_config.hf_config,
+                                  'final_logit_softcapping', None)
+        if logits_soft_cap is not None and self.attn_backend.get_name(
+        ) != "flashinfer":
+            logger.warning("Please use Flashinfer backend for models with"
+                           "logits_soft_cap (i.e., Gemma-2)."
+                           " Otherwise, the output might be wrong."
+                           " Set Flashinfer backend by "
+                           "export VLLM_ATTENTION_BACKEND=FLASHINFER.")
```
Comment (Yard1) [inline vllm/worker/model_runner.py:670]: we should just raise an exception IMO.

**Review by Yard1 — COMMENTED** [review]
we should check the flashinfer version and raise if it's too old

File: .buildkite/test-pipeline.yaml [hunk 2]
```diff
@@ -107,12 +107,15 @@ steps:
 
 - label: Kernels Test %N
   #mirror_hardwares: [amd]
-  command: pytest -v -s kernels --shard-id=$$BUILDKITE_PARALLEL_JOB --num-shards=$$BUILDKITE_PARALLEL_JOB_COUNT
+  command:
```
Comment (simon-mo) [inline .buildkite/test-pipeline.yaml:110]:
```suggestion
  commands:
```

**Review by Yard1 — APPROVED** [review]

File: vllm/worker/model_runner.py [hunk 3]
```diff
@@ -659,6 +659,16 @@ def _prepare_model_input_tensors(
                                            dtype=torch.long,
                                            device=self.device)
 
+        logits_soft_cap = getattr(self.model_config.hf_config,
+                                  'final_logit_softcapping', None)
+        if logits_soft_cap is not None and self.attn_backend.get_name(
+        ) != "flashinfer":
```
Comment (yongxb) [inline vllm/worker/model_runner.py:665]: Could I check if `logits_soft_cap` is supposed to be the `attn_logit_softcapping` value instead? The two values are different in the Gemma2 config.
```
"attn_logit_softcapping": 50.0,
"final_logit_softcapping": 30.0,
```

Comment (WoosukKwon) [reply vllm/worker/model_runner.py:665]: @yongxb Nice catch! `final_logit_softcapping` is used to cap the final logits before sampling. @LiuXiaoxuanPKU Could you please fix this?

**Review by comaniac — APPROVED** [review]

**Comment by zifeitong** [comment]
I think this warning can be removed to avoid confusion: https://github.com/vllm-project/vllm/blob/7cd2ebb0251fd1fd0eec5c93dac674603a22eddd/vllm/model_executor/models/gemma2.py#L140

**Comment by zifeitong** [comment]
I am able to run and reproduce the reported MMLU scores for both 9b and 27b models 👍 

However, if I don't disable CUDA graph, vLLM will crash with this error:

```
ERROR 07-04 03:34:04 async_llm_engine.py:53]   File "vllm/vllm/worker/model_runner.py", line 1202, in execute_model
ERROR 07-04 03:34:04 async_llm_engine.py:53]     model_input.attn_metadata.decode_wrapper = self.graph_runners[
ERROR 07-04 03:34:04 async_llm_engine.py:53] IndexError: list index out of range
```

**Comment by LiuXiaoxuanPKU** [comment]
> I am able to run and reproduce the reported MMLU scores for both 9b and 27b models 👍
> 
> However, if I don't disable CUDA graph, vLLM will crash with this error:
> 
> ```
> ERROR 07-04 03:34:04 async_llm_engine.py:53]   File "vllm/vllm/worker/model_runner.py", line 1202, in execute_model
> ERROR 07-04 03:34:04 async_llm_engine.py:53]     model_input.attn_metadata.decode_wrapper = self.graph_runners[
> ERROR 07-04 03:34:04 async_llm_engine.py:53] IndexError: list index out of range
> ```

Thanks for reporting! Could you give me an minimal reproducible example since I can run gemma-2 with flashinfer cudagraph on my end. Thanks!

**Comment by zifeitong** [comment]
I am using the `run_batch` script:

```
python -m vllm.entrypoints.openai.run_batch -i requests.jsonl -o /dev/null --model google/gemma-2-9b-it --disable-log-request
```

[requests.jsonl.zip](https://github.com/user-attachments/files/16090977/requests.jsonl.zip)

**Comment by LiuXiaoxuanPKU** [comment]
> I am using the `run_batch` script:
> 
> ```
> python -m vllm.entrypoints.openai.run_batch -i requests.jsonl -o /dev/null --model google/gemma-2-9b-it --disable-log-request
> ```
> 
> [requests.jsonl.zip](https://github.com/user-attachments/files/16090977/requests.jsonl.zip)

I tried the script and data on H100, it seems work. Could you report your environment? Flashinfer only supports GPU with compute capability greater than 8.0 (https://developer.nvidia.com/cuda-gpus). Not sure if that might be the problem.

**Comment by zifeitong** [comment]
> > I am using the `run_batch` script:
> > ```
> > python -m vllm.entrypoints.openai.run_batch -i requests.jsonl -o /dev/null --model google/gemma-2-9b-it --disable-log-request
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [requests.jsonl.zip](https://github.com/user-attachments/files/16090977/requests.jsonl.zip)
> 
> I tried the script and data on H100, it seems work. Could you report your environment? Flashinfer only supports GPU with compute capability greater than 8.0 (https://developer.nvidia.com/cuda-gpus). Not sure if that might be the problem.

I am using H100 with CUDA 12.5. Can you try sync you branch to the latest? https://github.com/vllm-project/vllm/pull/4412 might be related (it refactors `graph_runners`).

**Comment by LiuXiaoxuanPKU** [comment]
> > > I am using the `run_batch` script:
> > > ```
> > > python -m vllm.entrypoints.openai.run_batch -i requests.jsonl -o /dev/null --model google/gemma-2-9b-it --disable-log-request
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > [requests.jsonl.zip](https://github.com/user-attachments/files/16090977/requests.jsonl.zip)
> > 
> > 
> > I tried the script and data on H100, it seems work. Could you report your environment? Flashinfer only supports GPU with compute capability greater than 8.0 (https://developer.nvidia.com/cuda-gpus). Not sure if that might be the problem.
> 
> I am using H100 with CUDA 12.5. Can you try sync you branch to the latest? #4412 might be related (it refactors `graph_runners`).

yes, it's a merge conflict. Just fixed, please try again. Thanks!

**Comment by zifeitong** [comment]
> > > > I am using the `run_batch` script:
> > > > ```
> > > > python -m vllm.entrypoints.openai.run_batch -i requests.jsonl -o /dev/null --model google/gemma-2-9b-it --disable-log-request
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > [requests.jsonl.zip](https://github.com/user-attachments/files/16090977/requests.jsonl.zip)
> > > 
> > > 
> > > I tried the script and data on H100, it seems work. Could you report your environment? Flashinfer only supports GPU with compute capability greater than 8.0 (https://developer.nvidia.com/cuda-gpus). Not sure if that might be the problem.
> > 
> > 
> > I am using H100 with CUDA 12.5. Can you try sync you branch to the latest? #4412 might be related (it refactors `graph_runners`).
> 
> yes, it's a merge conflict. Just fixed, please try again. Thanks!

Thanks for the fix. It works now, w/ or w/o CUDA graph.

**Pull request merged**